### PR TITLE
✨ Add Proxy and SSL setting

### PR DESCRIPTION
**Pull request by dugulingping**
Hi I added some new functional methods, I hope to merge to the Master branch.
- Add Proxy support
- Add SSL support, Now able to make normal SSL

### CODE REVIEW

File: src/Curl/Curl.php [hunk 1]
```diff
@@ -571,6 +571,46 @@ public function setVerbose($on = true)
...
+     * @param   bool $type    When the param is set to FALSE, the certificate check and authentication will be ignored.
+     * @return  self
+     * @author  dugulingping
+     * @email   dugulingping@outlook.com
+     */
+    public function setSsl($type = true) : Curl{
+        $this->setOpt(CURLOPT_SSL_VERIFYPEER, $type);
+        // SSL CA pubilc key file ： http://curl.haxx.se/ca/cacert.pem
+        // After downloading, please put it in the folder where the library is located
+        $this->setOpt(CURLOPT_CAINFO, dirname(__FILE__).'/cacert.pem');
```
Comment (amouhzi) [inline src/Curl/Curl.php:610]: I think we should pass it as an argument.

File: src/Curl/Curl.php [hunk 2]
```diff
@@ -571,6 +571,46 @@ public function setVerbose($on = true)
...
+     *
+     * @param   string $type      value can be CURLPROXY_HTTP(default),CURLPROXY_SOCKS4,CURLPROXY_SOCKS5,CURLPROXY_SOCKS4A and CURLPROXY_SOCKS5_HOSTNAME
+     * @param   int $port         default: 8080
+     * @param   string $host      default: 127.0.0.1
+     * @param   string $userpwd   default: Null, example: 'username.password', value are string
+     * @return  self
+     * @author  dugulingping
+     * @email   dugulingping@outlook.com
+     */
+    public function setProxy($type = 'CURLPROXY_HTTP',$port = 8080, $host = '127.0.0.1', $userpwd = null) : Curl{
```
Comment (amouhzi) [inline src/Curl/Curl.php:586]: It is usual to put the host before the port.

What about username ?

File: src/Curl/Curl.php (hunk 1 above)
Comment (dugulingping) [reply src/Curl/Curl.php]: OK，I have modified it.

File: src/Curl/Curl.php (hunk 2 above)
Comment (dugulingping) [reply src/Curl/Curl.php]: Now $host Before $port 

Now, The Username and Password at $userAndPasswd, They are merger, The format is: 'username.password'

**Comment by dugulingping** [comment]
The new 'cacert.pem' file is public CA Root Certificates, not private SSH/SSL key.

File: src/Curl/Curl.php [hunk 3]
```diff
@@ -571,6 +571,54 @@ public function setVerbose($on = true)
...
+     * @email   dugulingping@outlook.com
+     */
+    public function setProxy($type = 'CURLPROXY_HTTP', $host = '127.0.0.1', $port = 8080,  $userAndPasswd = null) : Curl{
+        $this->setOpt(CURLOPT_PROXYAUTH, 'CURLAUTH_BASIC');
+        $this->setOpt(CURLOPT_PROXYTYPE, $type);
+        $this->setOpt(CURLOPT_PROXY, $host);
+        $this->setOpt(CURLOPT_PROXYPORT, $port);
+        if(isset($userpwd)) {
+            //The format is: 'username.password'
+            $this->setOpt(CURLOPT_PROXYUSERPWD, $userAndPasswd);
```
Comment (amouhzi) [inline src/Curl/Curl.php:593]: The format is not as you mentioned.

CURLOPT_PROXYUSERPWD | A username and password formatted as "[username]:[password]" to use for the connection to the proxy.

I prefer to have two variables: **$username** and **$password**. I think it's clearer like that.

File: src/Curl/Curl.php [hunk 4]
```diff
@@ -571,6 +571,54 @@ public function setVerbose($on = true)
...
+     * set SSL
+     *
+     * @param   bool $type    When the param is set to FALSE, the certificate check and authentication will be ignored.
+     * @param   string $filePath
+     * @param   string $fileName
+     * @return  self
+     * @author  dugulingping
+     * @email   dugulingping@outlook.com
+     */
+    public function setSsl($type = true, $filePath = '', $fileName =  'cacert.pem') : Curl{
```
Comment (amouhzi) [inline src/Curl/Curl.php:609]: We don't need two variables here, only one is enough, the **$filepath** variable can contain the full path of the file.

File: src/Curl/Curl.php (hunk 3 above)
Comment (dugulingping) [reply src/Curl/Curl.php]: Oh, Sorrry, Now, I've modified it.

File: src/Curl/Curl.php (hunk 4 above)
Comment (dugulingping) [reply src/Curl/Curl.php]: Ok, modified it.

**Comment by nadar** [comment]
@amouhzi is there something missing?

**Comment by nadar** [comment]
@dugulingping i would love to merge this, but i would say we need a UnitTest. Unless this happens:

---

I am going to close this issue to maintain a clean issue tracker for the following reasons:

1. Incomplete information (Are there any unanswered questions? Please provide as much information as possible).
2. Lack of recent activity (No further replies).
3. No one has taken up the task (no Pull Request has been received).

If you believe that this issue remains important or if you have additional or updated information, please consider reopening the issue and informing us.

**Pull request closed without merging**